ingress/elb/elb_test.go

```diff
@@ -0,0 +1,271 @@
+		Cluster: "test",
+	})
+
+	assert.EqualError(t, err, "unable to describe load balancers: oh dear oh dear")
+}
+
+func TestErrorDescribingTags(t *testing.T) {
+	e, mockElb, mockMetadata := setup()
+	instanceID := "cow"
+	mockInstanceMetadata(mockMetadata, instanceID)
+	mockLoadBalancers(mockElb, "one")
+	mockElb.On("DescribeTags", mock.AnythingOfType("*elb.DescribeTagsInput")).Return(&aws_elb.DescribeTagsOutput{}, errors.New("oh dear oh dear"))
+
+	_, err := e.Attach(types.FrontendInput{Cluster: "test"})
+
+	assert.EqualError(t, err, "unable to describe tags: oh dear oh dear")
+}
+
+func TestNoMatchingElbs(t *testing.T) {
+	// given
+	e, mockElb, mockMetadata := setup()
+	instanceID := "cow"
+	loadBalancerName := "i am not the loadbalancer you are looking for"
+	mockInstanceMetadata(mockMetadata, instanceID)
+	mockLoadBalancers(mockElb, loadBalancerName)
+	// No cluster tags
+	mockClusterTags(mockElb, lbTags{name: loadBalancerName, tags: []*aws_elb.Tag{}})
+
+	// when
+	number, err := e.Attach(types.FrontendInput{Cluster: clusterName})
+
+	// then
+	assert.NoError(t, err)
+	assert.Equal(t, 0, number)
+}
+
+func TestGetLoadBalancerPages(t *testing.T) {
+	// given
+	e, mockElb, mockMetadata := setup()
+	instanceID := "cow"
+	loadBalancerName := "lb1"
+	mockElb.On("DescribeLoadBalancers", &aws_elb.DescribeLoadBalancersInput{}).Return(&aws_elb.DescribeLoadBalancersOutput{NextMarker: aws.String("Use me")}, nil)
+	mockElb.On("DescribeLoadBalancers", &aws_elb.DescribeLoadBalancersInput{Marker: aws.String("Use me")}).Return(&aws_elb.DescribeLoadBalancersOutput{
+		LoadBalancerDescriptions: []*aws_elb.LoadBalancerDescription{&aws_elb.LoadBalancerDescription{LoadBalancerName: aws.String(loadBalancerName)}},
+	}, nil)
+	mockInstanceMetadata(mockMetadata, instanceID)
+	mockClusterTags(mockElb, lbTags{name: loadBalancerName, tags: []*aws_elb.Tag{&aws_elb.Tag{Key: aws.String(frontendTag), Value: aws.String(clusterName)}}})
+	mockRegisterInstances(mockElb, loadBalancerName, instanceID)
+
+	// when
+	number, err := e.Attach(types.FrontendInput{Cluster: clusterName})
+
+	// then
+	assert.Equal(t, 1, number)
+	assert.NoError(t, err)
+	mock.AssertExpectationsForObjects(t, mockElb.Mock)
+}
+
+func TestTagCallsPage(t *testing.T) {
+	// given
+	e, mockElb, mockMetadata := setup()
+	e.(*elb).maxTagQuery = 1
+	instanceID := "cow"
+	loadBalancerName1 := "lb1"
+	loadBalancerName2 := "lb1"
+	mockInstanceMetadata(mockMetadata, instanceID)
+	mockLoadBalancers(mockElb, loadBalancerName1, loadBalancerName2)
+	mockClusterTags(mockElb, lbTags{name: loadBalancerName1, tags: []*aws_elb.Tag{&aws_elb.Tag{Key: aws.String(frontendTag), Value: aws.String(clusterName)}}})
+	mockClusterTags(mockElb, lbTags{name: loadBalancerName2, tags: []*aws_elb.Tag{&aws_elb.Tag{Key: aws.String(frontendTag), Value: aws.String(clusterName)}}})
+	mockRegisterInstances(mockElb, loadBalancerName1, instanceID)
+	mockRegisterInstances(mockElb, loadBalancerName2, instanceID)
```

Comment:
So many mocks! This scares me for future maintenance. But this class is the abstraction on AWS, so I see why we need to do it to fully test `elbs` behaviour. I guess in the future if the logic becomes more complex (such as with dynamic ELB updating) we may want to extract the AWS calls into a thin wrapper that we can mock instead.